### PR TITLE
Support defaultOverscrollFactory() on iOS

### PR DESCRIPTION
**Pull request by ASalavei**
Add implementation of OverscrollFactory to support new Overscroll API.
Fix overscroll in CoreTextField and BasicTextField.
Upstreaming issue added: https://youtrack.jetbrains.com/issue/CMP-7307/Upstreaming.feature.foundation.text.rememberTextFieldOverscrollEffect

Fixes https://youtrack.jetbrains.com/issue/CMP-7143/Support-OverscrollFactory-and-LocalOverscrollFactory

### Fixes - iOS
- Enables Cupertino Overscroll by default for scrollable components
- Experimental method `optOutOfCupertinoOverscroll()` removed.

### CODE REVIEW

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/gestures/UikitScrollable.uikit.kt [hunk 1]
```diff
@@ -14,37 +14,19 @@
...
 internal actual fun CompositionLocalConsumerModifierNode.platformScrollConfig(): ScrollConfig = UiKitScrollConfig
 
-/**
- * Opt out of the Cupertino overscroll behavior (rubber banding and spring effect).
- *
- * This method should be called before any @Composable function using this effect is executed
- * (so as early as possible, e.g. during app start up).
- */
-@ExperimentalFoundationApi
-fun optOutOfCupertinoOverscroll() {
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/gestures/UikitScrollable.uikit.kt:41]: Please mention removing this in the release notes

File: compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/text/BasicTextField.kt [hunk 2]
```diff
@@ -313,6 +314,8 @@ internal fun BasicTextField(
 
     DisposableEffect(textFieldSelectionState) { onDispose { textFieldSelectionState.dispose() } }
 
+    val overscrollEffect = rememberTextFieldOverscrollEffect()
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/text/BasicTextField.kt:317]: Is it a regression after merging? Why should it be in the same PR? Why isn't it AOSP first edit?

Comment (ASalavei) [reply compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/text/BasicTextField.kt:317]: > Is it a regression after merging?

Yes

> Why should it be in the same PR?

It fixes bugs within the overscroll functionality. Basically, keeps the code that was added by us and removed then after AOSP merge.  If you feel that it should be in another one, please let me know the reasons.

> Why isn't it AOSP first edit?

1. The feature has not beed upstreamed yet.
2. Too long feedback loop.

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/cupertino/CupertinoOverscrollEffect.kt [hunk 3]
```diff
@@ -148,8 +151,8 @@ class CupertinoOverscrollEffect(
 
     private fun NestedScrollSource.toCupertinoScrollSource(): CupertinoScrollSource? =
         when (this) {
-            NestedScrollSource.Drag -> CupertinoScrollSource.DRAG
-            NestedScrollSource.Fling -> CupertinoScrollSource.FLING
```
Comment (mazunin-v-jb) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/cupertino/CupertinoOverscrollEffect.kt:152]: Why was it renamed? I might be slightly out of context here, could you please explain?

Comment (ASalavei) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/cupertino/CupertinoOverscrollEffect.kt:152]: Sure, the old values where deprecated and no longer supported.

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/gestures/UikitScrollable.uikit.kt (hunk 1 above)
Comment (ASalavei) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/gestures/UikitScrollable.uikit.kt:41]: Done

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Overscroll.uikit.kt [hunk 4]
```diff
@@ -17,32 +17,56 @@
...
-    null
+        if (density != other.density) return false
+        if (layoutDirection != other.layoutDirection) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        var result = density.hashCode()
+        result = 31 * result + layoutDirection.hashCode()
```
Comment (MatkovIvan) [inline compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Overscroll.uikit.kt]: jfyi data classes are forbidden only for public API

**Review by MatkovIvan — APPROVED** [review]

File: compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Overscroll.uikit.kt (hunk 4 above)
Comment (ASalavei) [reply compose/foundation/foundation/src/uikitMain/kotlin/androidx/compose/foundation/Overscroll.uikit.kt]: Makes sense. Updated

**Review by mazunin-v-jb — APPROVED** [review]

**Pull request merged**